Upload states task

alexboliam opened the pull request.

alexboliam committed:
Upload states task

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/bloc.dart:
```diff
@@ -0,0 +1,29 @@
+  final _inputController = BehaviorSubject<RecoveryPasswordState>();
+  Stream<RecoveryPasswordState> get stateStream => _inputController.stream;
+
+  void add({String email, bool isSumbitted}) {
```
Замените {String email, bool isSumbitted} на два Event-a:
RecoveryPasswordEmailChanged,
RecoveryPasswordSubmitted.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/bloc.dart:
```diff
@@ -0,0 +1,29 @@
+  Stream<RecoveryPasswordState> get stateStream => _inputController.stream;
+
+  void add({String email, bool isSumbitted}) {
+    if (_inputController.value == null)
```
У BehaviorSubject есть именованный конструктор (BehaviorSubjectю.seed() кажется). Туда можно сразу засунуть дефолтный стейт с пустым эмейлом и  isSubmitted = false. Тогда поулчится избавитья от этого ифа.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/bloc.dart:
```diff
@@ -0,0 +1,29 @@
+}
+
+class RecoveryPasswordState {
+  RecoveryPasswordState(this.email, this.isSumbitted);
```
Добавьте еще именованный конструктор initial(). Его можно будет использовать в BehaviorSubjectю.seed().

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -1,8 +1,43 @@
+  void initState() {
+    super.initState();
+
+    _emailController.addListener(() {
```
От этого можно избавиться. У TextField есть поле onChanged. Туда можно вставить _block.add(email: _emailController.text);

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -1,8 +1,43 @@
+      _block.add(email: _emailController.text);
+    });
+
+    _block.stateStream.listen(
```
Тут хорошо. Сделайте следующее:
1. В диалоговом окне 2 кнопки, "ок" и "cancel".
2. Нажимаю на кнопку и вовзращается из диалогового окна текст нажатой кнопки.
3. Перед строкой _block.add(isSumbitted: false); сделать принт текста нажатой кнопки в диалоговом окне. При чем принт должн быть сразу после метда showDialog.

(Это типичная задача, надо чтобы вы поняли как это делать)

idle-dt replied:
<img width="425" alt="зображення" src="https://user-images.githubusercontent.com/21138887/106350341-7f6bd580-62dd-11eb-9bb1-81e27c4d099c.png">
Вот тут хочу, чтобы была строка с принтом.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -1,8 +1,43 @@
+
+class _RecoveryPasswordState extends State<RecoveryPassword> {
+  final _emailController = TextEditingController();
+  final _block = RecoveryPasswordBloc();
```
Здесь вы создаете объект RecoveryPasswordBloc, но не вызываете у него dispose для освобождения ресурсов. Скорее всего, когда пользователь уйдет с этого экрана, то не удалится из памяти и этот экран, и блок, и стрим. С каждым заходом на этот экран будет создаваться новый объект блока, который не будет удаляться. 

У State<T> есть метод dispose. Там целесообразно вызвать _bloc.dispose().

(Если я этого не показал в примере, мой косяк)

KocHeto committed:
Fixes ver. 1.0

KocHeto committed:
Add block + code generation implementation

KocHeto committed:
Merge branch 'master' into f/recovery-password-state-task

idle-dt commented on .gitignore:
```diff
@@ -22,16 +22,16 @@
 
 # Flutter/Dart/Pub related
 **/doc/api/
+**/doc/api/
```
А это зачем ?)

idle-dt commented on .gitignore:
```diff
@@ -22,16 +22,16 @@
 .dart_tool/
 .flutter-plugins
+.flutter-plugins
+.flutter-plugins-dependencies
```
И эти две строки тоже. Дублирование получается.

idle-dt commented on .packages:
```diff
@@ -0,0 +1,75 @@
+# Generated by pub on 2021-02-14 21:08:36.126770.
```
А попробуй добавить этот файл в gitignore. Мне кажется, он лишний.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/bloc.dart:
```diff
@@ -0,0 +1,26 @@
+      : super(RecoveryPasswordState(
+            email: inputData.email, isSumbitted: inputData.isSumbitted));
+
+  RecoveryPasswordBloc.initial()
```
Я думаю, тебе не нужен в данном случае конструктор initial у блока. У него вообще должен быть один конструктор, по сути.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -1,8 +1,47 @@
+
   @override
   Widget build(BuildContext context) => Container(
         child: SafeArea(
```
Начнем. 
1.  Самый первый Container надо обвернуть в Scaffold, а его, в свою очередь, в BlocProvider. У него есть аргумент-метод create, в нем надо создать сам блок. Тогда код в dispose можно будет удалить, как и само переопределение.
2. Ту часть с listen, которая у тебя в initState, надо запихнуть в виджет BlocListener. Его можешь поставить сразу после Scaffold-a.

По хорошему, дерево у тебя будет выглядеть след. образом:
1. BlocProvider
2. Scaffold
3. BlocListener (в аргументе listen указываешь все, что нужно сделать при определенном состоянии. Всю часть, которая у тебя в initState -> listen можешь перенести сюда).
4. Container

KocHeto committed:
Fix of previous commit

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,114 @@
+                BlocListener<RecoveryPasswordBloc, RecoveryPasswordState>(
+              listener: (context, state) {
+                if (state.isSumbitted == true) {
+                  showDialog(
```
А можете вынести это в какой-то метод виджета?

К примеру, showConfirmationDialog с передачей в него контекста?

idle-dt replied:
Мне кажется, так чище будет выглядеть код.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,114 @@
+                      content: Text('You typed: ${state.email}'),
+                    ),
+                  ).then(print);
+                  _block.add(const RecoveryEvent.reset());
```
Это сильно, конечно, через than, но лучше используйте в таких ситуациях await. Оно читается комфортнее ))

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,114 @@
+                              _block.add(const RecoveryEvent.submit());
+                            },
+                            child: const Padding(
+                              padding: EdgeInsets.fromLTRB(21, 9, 21, 9),
```
В таких ситуациях лучше использовать EdgeInsets.symethric(vertical: 9, horizontal: 21)

KocHeto committed:
Add function showConfirmationDialog and make async print of modalWindow state

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,120 @@
+
+  @override
+  Widget build(BuildContext context) => BlocProvider<RecoveryPasswordBloc>(
+        create: (context) => _block,
```
Та прямо тут инстанс создавай)

idle-dt replied:
Учитывая, что у тебя StatelessWidget, этот метод вызовется 1 раз.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,120 @@
+        ),
+      );
+  void showConfirmationDialog(
+      BuildContext context, RecoveryPasswordState state) async {
```
Будет хорошо, если убрать от сюда стейт и положить isLoading и email

KocHeto committed:
Fix of previous commit

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,123 @@
+  @override
+  Widget build(BuildContext context) => BlocProvider<RecoveryPasswordBloc>(
+        create: (context) => RecoveryPasswordBloc(
+            const RecoveryPasswordState(email: '', isSubmitted: false)),
```
Смотри, вот это надо убрать из вьюхи и внутри блока сделать так:

RecoveryPasswordBloc(): super(RecoveryPasswordState.initial());

А уже в RecoveryPasswordState.initial() сделать дефолтные поля. Если бы у тебя были входные данные в блок, то каждый аргумент передавался бы в конструктор блока по отдельности.

Очень важно. Только блок имеет право создавать новый стейт!

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,123 @@
+        create: (context) => RecoveryPasswordBloc(
+            const RecoveryPasswordState(email: '', isSubmitted: false)),
+        child: Scaffold(
+          body: Builder(
```
Убираем билдер.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,123 @@
+                          style: TextStyle(fontSize: 30),
+                        ),
+                        const SizedBox(height: 10),
+                        TextField(
```
Обворачиваем TextField в билдер.

idle-dt commented on lib/examples_from_courses/presentation/pages/recovery_password/recovery_password.dart:
```diff
@@ -0,0 +1,123 @@
+                        const SizedBox(height: 60),
+                        Align(
+                          alignment: Alignment.center,
+                          child: RaisedButton(
```
Кнопку тоже обворачиваем в билдер.

idle-dt reviewed (commented):
Я думаю, тебе надо сделать последние правки и все)

Почему надо обворачивать каждый виджет по отдельности в билдер, а не какого-то общего родителя?
Потом, если надо будет опправить вьюху, заколебешься искать этот билдер.

KocHeto committed:
- Add RecoveryPasswordState.initial constuctor
- Add base inicializtion of RecoveryPasswordBlock
- Move Builder from listener to each element, that uses context from it
